src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackSerializer.cs

```diff
@@ -462,58 +467,171 @@ private static int LZ4Operation(in ReadOnlySequence<byte> input, Span<byte> outp
-                    var extReader = new MessagePackReader(extension.Data);
-
-                    // The first part of the extension payload is a MessagePack-encoded Int32 that
-                    // tells us the length the data will be AFTER decompression.
-                    int uncompressedLength = extReader.ReadInt32();
-
-                    // The rest of the payload is the compressed data itself.
-                    ReadOnlySequence<byte> compressedData = extReader.Sequence.Slice(extReader.Position);
-
-                    Span<byte> uncompressedSpan = writer.GetSpan(uncompressedLength).Slice(0, uncompressedLength);
-                    int actualUncompressedLength = LZ4Operation(compressedData, uncompressedSpan, LZ4Codec.Decode);
-                    Debug.Assert(actualUncompressedLength == uncompressedLength, "Unexpected length of uncompressed data.");
-                    writer.Advance(actualUncompressedLength);
-                    return true;
+                    MessagePackReader peekReader = reader.CreatePeekReader();
+                    ExtensionHeader header = peekReader.ReadExtensionFormatHeader();
+                    if (header.TypeCode == ThisLibraryExtensionTypeCodes.LZ4)
+                    {
+                        // Read the extension using the original reader, so we "consume" it.
+                        ExtensionResult extension = reader.ReadExtensionFormat();
+                        var extReader = new MessagePackReader(extension.Data);
+
+                        // The first part of the extension payload is a MessagePack-encoded Int32 that
+                        // tells us the length the data will be AFTER decompression.
+                        int uncompressedLength = extReader.ReadInt32();
+
+                        // The rest of the payload is the compressed data itself.
+                        ReadOnlySequence<byte> compressedData = extReader.Sequence.Slice(extReader.Position);
+
+                        Span<byte> uncompressedSpan = writer.GetSpan(uncompressedLength).Slice(0, uncompressedLength);
+                        int actualUncompressedLength = LZ4Operation(compressedData, uncompressedSpan, LZ4CodecDecode);
+                        Debug.Assert(actualUncompressedLength == uncompressedLength, "Unexpected length of uncompressed data.");
+                        writer.Advance(actualUncompressedLength);
+                        return true;
+                    }
+                }
+
+                // Try to find LZ4ContiguousBlock
+                if (reader.NextMessagePackType == MessagePackType.Array)
+                {
+                    MessagePackReader peekReader = reader.CreatePeekReader();
+                    var arrayLength = peekReader.ReadArrayHeader();
+                    if (arrayLength != 0 && peekReader.NextMessagePackType == MessagePackType.Extension)
+                    {
+                        ExtensionHeader header = peekReader.ReadExtensionFormatHeader();
```

Comment:
At some point you should switch back from `peekReader` to just the `reader`, or perhaps "fast forward" the `reader` at the end. The important bit being that you "consume" the bytes from the original reader as the `Lz4Block` section above this new one does.

Reply:
Thank you, right.